### ISSUE TITLE
[Enhancement]: Make some of the new checkout shortcode changes filterable

### ISSUE THREAD

**Issue by colmtroy** [woothemes/woocommerce#40079]
### Describe the solution you'd like

7.9.0 added the following checks for non-guest orders to the checkout shortcode

https://github.com/woocommerce/woocommerce/blob/60fa13e21b385f78b77d400f1b965ade2cf2f013/plugins/woocommerce/includes/shortcodes/class-wc-shortcode-checkout.php#L302

```
		$order_customer_id = $order->get_customer_id();

		// For non-guest orders, require the user to be logged in before showing this page.
		if ( $order_customer_id && get_current_user_id() !== $order_customer_id ) {
			wc_get_template( 'checkout/order-received.php', array( 'order' => false ) );
			wc_print_notice( esc_html__( 'Please log in to your account to view this order.', 'woocommerce' ), 'notice' );
			woocommerce_login_form( array( 'redirect' => $order->get_checkout_order_received_url() ) );
			return;
		}
```

This has had a big impact on lots of WooCommerce users - some of whom have had no choice but to start editing core:
https://wordpress.org/support/topic/email-verification-after-woo-order-from-guest/page/3/#post-17027691

On my own store I have a custom function to allow existing customers place an order without logging in (i assign the order to their account based on the billing email so they're no guest checkouts. I also disallow guest checkouts). 

My use case means customers end up seeing a login screen rather than the order thank you page.

I'd propose wrapping a filter around this new code added in 7.9.0 to allow merchants who need to disable it to do easily and safely without having to edit core, or do other dangerous things like overriding the user id etc.

**Comment by njeffers** [comment]
I second this. This is a **MAJOR** issue for us, as we allow guest checkout. It's just super confusing, and overall a bad shopping experience for you to place an order and get told to login just to view your order information. _How can they login if they don't have an account yet?!_ I've never seen it before when purchasing anything online. I've been through the forums, and tried using the 'woocommerce_order_email_verification_required' filter without luck. I've disabled all but WC, and activated 2023 - no dice.

I'm all for the added security, and pushing in that direction, but always felt the order_key you guys pass as a query parameter was plenty, and it's been that way since I can remember. I don't see how someone could get ahold of that link to the see address info, outside of someone sharing it purposefully, and I don't think you can edit anything from there anyway, so I don't see the real concern.

In spirit of pushing improved security, can we not just add a quick filter at the beginning of guest_should_verify_email() that allows us to bypass this?

**Comment by barryhughes** [comment]
1. This seems like it is a duplicate of https://github.com/woocommerce/woocommerce/issues/39750, so I'll go ahead and close it. If you feel I've misunderstood, please let me know and we'll be happy to reassess.
2. "How can they login if they don't have an account yet?" I'm not exactly sure what you were seeing, but just to emphasize: so far as I know, we only requested they login if the order was associated with a user account. Otherwise, we asked that they confirm the email address (which is not the same as logging in).
3. It should be possible to disable this new check via a filter hook (see snippet below), but it turns out there are conditions where that does not seem to work. https://github.com/woocommerce/woocommerce/pull/40050 is one part of the solution for this, and should be available in our 8.2.0 release.

```php
add_filter( 'woocommerce_order_email_verification_required', '__return_false' );
```
